### ISSUE TITLE
Enable Full Telemetry for users who enroll in Test Pilot

### ISSUE THREAD

**Issue by javaun** [mozilla/testpilot#655]
TL;DR tasks:
- [x] Preserve initial values of `toolkit.telemetry.enabled` and `datareporting.healthreport.uploadEnabled` preferences.
- [x] Set those two prefs to `true` on main add-on install / enable
- [x] Restore initial values of those two prefs on main add-on uninstall / disable

---

To enable us to answer behavioral questions about how users interact w/ experiments and whether they actually change browsing behavior, we want to enroll all Test Pilot users in full telemetry (opt-out + opt-in). 

Some measure of this is required for v1 launch.

**ENG Requirements**
- Full data: Test Pilot users send full Telemetry dataset to Telemetry system. (TBD on whether we just flip opt-in Telemetry on, or whether we send this data some other way, such as stuffing it into the custom Test Pilot Ping)
- Un-enrolling: If a user completely leaves Test Pilot, we need to restore their Telemetry to the prior state.
- Privacy needs: Telemetry ID cannot be mapped to their FxA account. (Illustrative purposes only: This might be achieved multiple ways, for example via siloed data storage, or by injecting some new identifier in place of the Telemetry ID, i.e. a one-way hash) 

**Corner cases:**
- If a user uninstalls Test Pilot via about:add-ons, we still need to clean up their Telemetry settings (if possible). 
- At launch, if a user manually turns off Telemetry (opt-out, opt-in, etc.), Test Pilot still works, Mozilla just doesn’t get data. (specifying this as a test case only, since I assume this is the current behavior. If this is widely abused, we may have to build a feature to notify users that Test Pilot requires telemetry)

**Comment by jaredhirsch** [comment]
This should be pretty easy to do. We can do stuff when the add-on is being installed/enabled and uninstalled/disabled. The simplest approach is probably to record what the user's Telemetry preferences were before joining Test Pilot, then alter the prefs as part of the add-on install process. When it's uninstalled, we can restore the original prefs.

**Comment by jaredhirsch** [comment]
Hmm, it [looks like](https://dxr.mozilla.org/mozilla-central/source/modules/libpref/init/all.js#851) a restart is required after enabling telemetry. Just something to be aware of.

**Comment by javaun** [comment]
We can wait until their next restart and not force them to alter their path. The only thing we lose is a bit of data for that session and that's preferable to interruption. The question is whether the user sees a prompt to restart or if we can suppress that.

**Comment by lmorchard** [comment]
> Full data: Test Pilot users send full Telemetry dataset to Telemetry system. (TBD on whether we just flip opt-in Telemetry on, or whether we send this data some other way, such as stuffing it into the custom Test Pilot Ping)

Test Pilot uses the Telemetry system to send pings. So, if I understand correctly, there's no Test Pilot ping without Telemetry enabled. So, I don't think that's a backup approach we can use, because it's all the same system.

> Privacy needs: Telemetry ID cannot be mapped to their FxA account. (Illustrative purposes only: This might be achieved multiple ways, for example via siloed data storage, or by injecting some new identifier in place of the Telemetry ID, i.e. a one-way hash) 

Is this a known issue right now? As far as I understand, we never associate a FxA ID with the Telemetry ID in any pings. If that's not the case, it's a bug & a surprise to me.

**Comment by lmorchard** [comment]
> we want to enroll all Test Pilot users in full telemetry (opt-out + opt-in)

Also, what all exactly does this entail? I'm not super familiar with what all Telemetry does. What are the "opt-out + opt-in" items? (i.e. does just flipping on the `toolkit.telemetry.unified` pref cover everything, or are there more granular flags we have to enable?)

We should also make sure [the metrics plan](https://github.com/mozilla/testpilot/blob/master/docs/README-METRICS.md) is updated, if this is a change in plans

**Comment by javaun** [comment]
I spoke w/ the team in #telemetry.  This is the pref needed to enable opt-in telemetry is:

toolkit.telemetry.enabled;true

This is on by default in pre-release and off by default in release. This gives us extra instrumentation.

The opt-out portion of telemetry (what was formerly called FHR) is what we absolutely need because our two special pings (Test Pilot Ping, and Test Pilot Experiment ping) ride on this. That is on by default in all channels. We may want to check that one as well on enrollment, but it's unlikely to be off. I'm working to confirm that pref, it may be: 

datareporting.healthreport.uploadEnabled;true

@clouserw may know

**Comment by javaun** [comment]
@lmorchard the requirement to "reset a user's opt-in Telemetry state" would be to store the user's state before we flip opt-in telemetry to true for them. So, if they have the default setting in release:

toolkit.telemetry.enabled;false 

We can store that (perhaps in another pref). Then we flip it to true when they join Test Pilot. If they leave test pilot, we flip it back to false.

**Comment by ghost** [comment]
> > ```
> > Full data: Test Pilot users send full Telemetry dataset to Telemetry system. (TBD on whether we just flip opt-in Telemetry on, or whether we send this data some other way, such as stuffing it into the custom Test Pilot Ping)
> > ```
> > 
> > Test Pilot uses the Telemetry system to send pings. So, if I understand correctly, there's no Test Pilot ping without Telemetry enabled. So, I don't think that's a backup approach we can use, because it's all the same system.

That's incorrect.  Confusingly, our pings will go through, regardless of the opt-in flag we're discussing here.

> > ```
> > Privacy needs: Telemetry ID cannot be mapped to their FxA account. (Illustrative purposes only: This might be achieved multiple ways, for example via siloed data storage, or by injecting some new identifier in place of the Telemetry ID, i.e. a one-way hash)
> > ```
> > 
> > Is this a known issue right now? As far as I understand, we never associate a FxA ID with the Telemetry ID in any pings. If that's not the case, it's a bug & a surprise to me.

This is not an issue now.  We avoid connecting the two on purpose.

**Comment by ghost** [comment]
We should ensure both telemetry reporting flags are on.  Thanks for finding what they are @javaun .

**Comment by lmorchard** [comment]
Cool. So, to sum up, it sounds like this is what needs doing:
- [ ] Preserve initial values of `toolkit.telemetry.enabled` and `datareporting.healthreport.uploadEnabled` preferences.
- [ ] Set those two prefs to `true` on main add-on install / enable
- [ ] Restore initial values of those two prefs on main add-on uninstall / disable

**Comment by ghost** [comment]
perfect :100:

**Comment by javaun** [comment]
yes, perfect
